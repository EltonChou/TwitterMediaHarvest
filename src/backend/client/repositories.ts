import { Sha256 } from '@aws-crypto/sha256-browser'
import type { CognitoIdentityCredentials } from '@aws-sdk/credential-provider-cognito-identity'
import { FetchHttpHandler, streamCollector } from '@aws-sdk/fetch-http-handler'
import { HttpRequest } from '@aws-sdk/protocol-http'
import { SignatureV4 } from '@aws-sdk/signature-v4'
import { CreateClientFailed, UpdateStatsFailed } from '@backend/errors'
import type { IStorageProxy } from '@libs/proxy'
import type { ClientInfo, V4Statistics } from '@schema'
import jws from 'jws'
import type { Storage } from 'webextension-polyfill'
import Browser from 'webextension-polyfill'
import { ClientInfoVO } from './valueObjects'

type UpdateInfo = Pick<ClientInfo, 'csrfToken' | 'syncedAt'>

const defaultInfo: Readonly<ClientInfo> = {
  uuid: 'uuid',
  csrfToken: 'token',
  syncedAt: 0,
  uninstallCode: 'code',
}

const isEmptyInfo = (record: Record<string, unknown>) =>
  record['uuid'] === defaultInfo.uuid ||
  record['csrfToken'] === defaultInfo.csrfToken ||
  record['syncedAt'] === defaultInfo.syncedAt ||
  record['uninstall'] === defaultInfo.uninstallCode

type ProviderOptions = {
  credentialProvider: Provider<CognitoIdentityCredentials>
  statsProvider: Provider<V4Statistics>
}

export interface IClientInfoRepository {
  getInfo(options?: ProviderOptions): Promise<ClientInfoVO>
  updateStats(options?: ProviderOptions): Promise<void>
  resetInfo(): Promise<void>
}

export class InfoSyncLock implements IProcessLock {
  private lockCriteria = 'InfoUpdateLock'
  private maxLockTime = 10 * 60 * 1000

  constructor(readonly storageArea: Storage.StorageArea) {}

  async isLocked(): Promise<boolean> {
    const record = await this.storageArea.get(this.lockCriteria)
    return Object.keys(record).includes(this.lockCriteria) && record[this.lockCriteria] - Date.now() >= this.maxLockTime
  }

  async release(): Promise<void> {
    await this.storageArea.remove(this.lockCriteria)
  }

  async acquire(): Promise<void> {
    await this.storageArea.set({ [this.lockCriteria]: Date.now() })
  }
}

export class ClientInfoRepository implements IClientInfoRepository {
  constructor(readonly storageArea: IStorageProxy<ClientInfo>, private defaultOptions: ProviderOptions) {}

  private async createClient(options: ProviderOptions): Promise<ClientTokenResponse> {
    const initStats = await options.statsProvider()
    const credential = await options.credentialProvider()

    const handler = new ClientApiHandler()
    const response = await handler.send(HttpMethod.Post, '/clients', {}, initStats, credential)

    if (response.statusCode === 201) {
      const body = JSON.parse(response.body)
      return body
    }

    throw new CreateClientFailed(
      response.statusCode,
      response.body?.message || JSON.stringify(response.body),
      response.headers
    )
  }

  private async getNewInfo(options?: Partial<ProviderOptions>): Promise<ClientInfoVO> {
    const clientTokenResponse: ClientTokenResponse = await this.createClient({ ...this.defaultOptions, ...options })
    const { payload } = jws.decode(clientTokenResponse.token, { json: true })
    const info = new ClientInfoVO({
      uuid: payload['uuid'],
      csrfToken: clientTokenResponse.token,
      syncedAt: Date.now(),
      uninstallCode: clientTokenResponse?.uninstallCode || payload['uninstallCode'],
    })
    await this.storageArea.setItem(info.props)
    await Browser.runtime.setUninstallURL(info.uninstallUrl)
    return info
  }

  async getInfo(options?: Partial<ProviderOptions>): Promise<ClientInfoVO> {
    const record = await this.storageArea.getItemByDefaults(defaultInfo)
    return isEmptyInfo(record) ? this.getNewInfo(options) : new ClientInfoVO(record)
  }

  async updateStats(options?: Partial<ProviderOptions>): Promise<void> {
    options = { ...this.defaultOptions, ...options }

    const { uuid, csrfToken } = await this.getInfo(options)
    const stats = await options.statsProvider()
    const credential = await options.credentialProvider()

    const handler = new ClientApiHandler()
    const response = await handler.send(
      HttpMethod.Put,
      '/clients/' + uuid + '/stats',
      { 'x-csrf-token': csrfToken },
      stats,
      credential
    )

    if (response.statusCode !== 200)
      throw new UpdateStatsFailed(
        response.statusCode,
        response.body?.message || JSON.stringify(response.body),
        response.headers
      )

    const body = JSON.parse(response.body)
    const clientInfo: UpdateInfo = {
      csrfToken: body.token,
      syncedAt: Date.now(),
    }
    await this.storageArea.setItem(clientInfo)
  }

  async resetInfo(): Promise<void> {
    await this.storageArea.setItem(defaultInfo)
  }
}

enum HttpMethod {
  Post = 'POST',
  Get = 'GET',
  Put = 'PUT',
  Delete = 'DELETE',
}

class ClientApiHandler {
  async send(
    method: HttpMethod,
    path: string,
    headers: Record<string, string>,
    body: unknown,
    credentials: CognitoIdentityCredentials
  ) {
    const request = new HttpRequest({
      protocol: 'https',
      body: JSON.stringify(body),
      headers: {
        'Content-Type': 'application/json',
        'X-Api-Key': process.env.API_KEY,
        'X-MediaHarvest-Version': process.env.VERSION_NAME,
        host: process.env.API_HOSTNAME,
        ...headers,
      },
      method: method,
      hostname: process.env.API_HOSTNAME,
      path: (process.env.API_ROOT_PATH || '') + path,
    })

    const signer = new SignatureV4({
      region: 'ap-northeast-1',
      service: 'execute-api',
      credentials: credentials,
      sha256: Sha256,
    })

    const signedRequest = await signer.sign(request)

    const client = new FetchHttpHandler()
    const { response } = await client.handle(signedRequest as HttpRequest)

    response.body = new TextDecoder().decode(await streamCollector(response.body))
    return response
  }
}
