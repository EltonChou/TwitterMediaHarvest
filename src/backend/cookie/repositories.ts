import browser from 'webextension-polyfill'

export interface ITwitterTokenRepository {
  getCsrfToken(): Promise<string | null>
  getGuestToken(): Promise<string | null>
}

/**
 * @deprecated
 */
export class TwitterTokenRepository implements ITwitterTokenRepository {
  async getCsrfToken(): Promise<string | null> {
    const cookie = await browser.cookies.get({
      url: 'https://twitter.com',
      name: 'ct0',
    })

    return cookie ? cookie.value : null
  }

  async getGuestToken(): Promise<string | null> {
    const cookie = await browser.cookies.get({
      url: 'https://twitter.com',
      name: 'gt',
    })
    return cookie ? cookie.value : null
  }
}

export class XTokenRepository implements ITwitterTokenRepository {
  async getCsrfToken(): Promise<string | null> {
    const cookie = await browser.cookies.get({
      url: 'https://x.com',
      name: 'ct0',
    })

    return cookie ? cookie.value : null
  }

  async getGuestToken(): Promise<string | null> {
    const cookie = await browser.cookies.get({
      url: 'https://x.com',
      name: 'gt',
    })
    return cookie ? cookie.value : null
  }
}
