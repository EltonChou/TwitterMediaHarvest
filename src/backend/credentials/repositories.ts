import type { CognitoIdentityCredentials, Storage } from '@aws-sdk/credential-provider-cognito-identity'
import { fromCognitoIdentityPool } from '@aws-sdk/credential-providers'
import ValueObject from '@backend/valueObject'
import type { AWSCredentials } from '@schema'
import type { Storage as BrowserStorage } from 'webextension-polyfill'
import Browser from 'webextension-polyfill'

class CredentialVO extends ValueObject<AWSCredentials> {
  readonly expireTimeRedundancy: number = 3 * 60 * 1000

  constructor(props: AWSCredentials) {
    super(props)
  }

  get credential(): CognitoIdentityCredentials {
    return { ...this.props, expiration: new Date(this.props.expiration) }
  }

  get isExpired(): boolean {
    return this.props?.expiration ? this.props.expiration < Date.now() - this.expireTimeRedundancy : true
  }

  static fromCognitoIdentityCredentials(credential: CognitoIdentityCredentials): CredentialVO {
    return new CredentialVO({
      ...credential,
      // ? Does it means the credential wouldn't expired if the expiration was undefined?
      expiration: credential?.expiration ? credential.expiration.getTime() : 0,
    })
  }
}

export interface ICredentialRepository {
  getCredential(): Promise<CognitoIdentityCredentials>
}

class BrowserSyncStorageProxy implements Storage {
  async getItem(key: string): Promise<string | null> {
    const record = await Browser.storage.sync.get(key)
    return Object.keys(record).length === 0 ? null : record[key]
  }

  async removeItem(key: string): Promise<void> {
    return await Browser.storage.sync.remove(key)
  }

  async setItem(key: string, data: string): Promise<void> {
    return await Browser.storage.sync.set({ [key]: data })
  }
}

export class CredentialRepository implements ICredentialRepository {
  readonly credentialCriteria = 'awsCredential'

  constructor(readonly storageArea: BrowserStorage.StorageArea) {}

  private async fetchCredential(): Promise<CognitoIdentityCredentials> {
    // ! Data in storage is `AWSCredential` not `CognitoIdentityCredentials`
    // TODO: Need to define the schema of storage area.
    const credential = await fromCognitoIdentityPool({
      identityPoolId: process.env.IDENTITY_POOL_ID,
      cache: new BrowserSyncStorageProxy(),
      clientConfig: {
        region: process.env.IDENTITY_POOL_REGION,
      },
    })()

    const credentialVO = CredentialVO.fromCognitoIdentityCredentials(credential)
    await this.storageArea.set({ [this.credentialCriteria]: credentialVO.props })
    return credential
  }

  async getCredential(): Promise<CognitoIdentityCredentials> {
    const credentialRecord = await this.storageArea.get(this.credentialCriteria)
    if (Object.keys(credentialRecord).length === 0) return await this.fetchCredential()

    const credentialsVO = new CredentialVO(credentialRecord[this.credentialCriteria])
    return credentialsVO.isExpired ? await this.fetchCredential() : credentialsVO.credential
  }
}
