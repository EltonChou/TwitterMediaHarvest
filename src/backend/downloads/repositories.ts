import { DownloadHistoryEntity, DownloadRecord } from './models'
import type { DownloadDBSchema, DownloadHistoryItem, HashtagItem } from '@schema'
import { type IDBPDatabase } from 'idb'
import type { Storage } from 'webextension-polyfill'

export interface IDownloadRecordsRepository {
  getById(downloadItemId: number): Promise<DownloadRecord | null>
  save(downloadRecord: DownloadRecord): Promise<void>
  removeById(downloadItemId: number): Promise<void>
  getAll(): Promise<DownloadRecord[]>
}

abstract class IndexedDBRepository {
  constructor(readonly clientProvider: Provider<IDBPDatabase<DownloadDBSchema>>) {}
}

export class IndexedDBDownloadRecordsRepository
  extends IndexedDBRepository
  implements IDownloadRecordsRepository
{
  async save(downloadRecord: DownloadRecord): Promise<void> {
    const client = await this.clientProvider()
    await client.put('record', downloadRecord.toJson())
  }

  async getById(downloadItemId: number): Promise<DownloadRecord | null> {
    const client = await this.clientProvider()
    const record = await client.get('record', downloadItemId)
    return record ? DownloadRecord.fromJson(record) : null
  }

  async removeById(downloadItemId: number): Promise<void> {
    const client = await this.clientProvider()
    await client.delete('record', downloadItemId)
  }

  async getAll(): Promise<DownloadRecord[]> {
    const client = await this.clientProvider()
    const records = await client.getAll('record')
    return records.map(DownloadRecord.fromJson)
  }
}

type DownloadRecordId = `dl_${number}`
const storageAreaDownloadIdPattern = /dl_(?:\d+)/

export class StorageAreaDownloadRecordsRepository implements IDownloadRecordsRepository {
  constructor(readonly storageArea: Storage.StorageArea) {}

  async save(downloadRecord: DownloadRecord): Promise<void> {
    await this.storageArea.set({
      [createId(downloadRecord.id)]: downloadRecord.toJson(),
    })
  }

  async getById(downloadItemId: number): Promise<DownloadRecord | null> {
    const recordId = createId(downloadItemId)
    const volume = await this.storageArea.get(recordId)
    if (recordId in volume) {
      return DownloadRecord.fromJson(volume[recordId])
    }
    return null
  }

  async removeById(downloadItemId: number): Promise<void> {
    const recordId = createId(downloadItemId)
    await this.storageArea.remove(recordId)
  }

  async getAll(): Promise<DownloadRecord[]> {
    const volumes = await this.storageArea.get(null)

    return Object.entries(volumes)
      .filter(([k, v]) => Boolean(k.match(storageAreaDownloadIdPattern)))
      .map(([k, v]) => DownloadRecord.fromJson(v))
  }
}

function createId(downloadItemId: number): DownloadRecordId {
  return `dl_${downloadItemId}`
}

export type DownloadItemPredicate = (item: DownloadHistoryItem) => boolean

export interface IDownloadHistoryRepository {
  count(): Promise<number>
  save(item: DownloadHistoryEntity): Promise<void>
  clear(): Promise<void>
  tweetHasDownloaded(tweetId: string): Promise<boolean>
  getAll(): Promise<DownloadHistoryEntity[]>
  getLatest(limit?: number): Promise<DownloadHistoryEntity[]>
  getByTweetId(tweetId: string): Promise<DownloadHistoryEntity | undefined>
  searchByUserName(userName: string, limit?: number): Promise<DownloadHistoryEntity[]>
  search(
    ...predicates: DownloadItemPredicate[]
  ): (limit?: number, skip?: number) => Promise<[DownloadHistoryEntity[], number]>
}

export class IndexedDBDownloadHistoryRepository
  extends IndexedDBRepository
  implements IDownloadHistoryRepository
{
  private openCursor(
    range?: IDBKeyRange,
    direction?: IDBCursorDirection,
    limit: number | 'inf' = 50,
    offset = 0
  ) {
    const transaction = this.transaction()
    const shouldContinue = (count: number) => (limit === 'inf' ? true : count < limit)

    return async function* (indexKey: keyof DownloadDBSchema['history']['indexes']) {
      let cursor = await (await transaction).store
        .index(indexKey)
        .openCursor(range, direction)

      if (offset) cursor = await cursor.advance(offset)
      let count = 0
      while (cursor && shouldContinue(count)) {
        yield cursor.value
        cursor = await cursor.continue()
        count += 1
      }
    }
  }

  async transaction(mode: IDBTransactionMode = 'readonly') {
    const client = await this.clientProvider()
    const tx = client.transaction('history', mode)
    return tx
  }

  async tweetHasDownloaded(tweetId: string): Promise<boolean> {
    const client = await this.clientProvider()
    const key = await client.getKey('history', tweetId)
    return Boolean(key)
  }

  async clear(): Promise<void> {
    const client = await this.clientProvider()
    await client.clear('history')
  }

  async save(item: DownloadHistoryEntity): Promise<void> {
    const client = await this.clientProvider()
    await client.put('history', item.toDownloadHistoryItem())
  }

  async count(): Promise<number> {
    const client = await this.clientProvider()
    return client.count('history')
  }

  async getAll(): Promise<DownloadHistoryEntity[]> {
    const client = await this.clientProvider()
    const items = await client.getAll('history')
    return items.map(DownloadHistoryEntity.build)
  }

  async getLatest(limit = 10, offset = 0): Promise<DownloadHistoryEntity[]> {
    const collection = this.openCursor(null, 'prev', limit, offset)('byDownloadTime')
    return (await asyncGeneratorToArray(collection)).map(DownloadHistoryEntity.build)
  }

  async getByTweetId(tweetId: string): Promise<DownloadHistoryEntity | undefined> {
    const client = await this.clientProvider()
    const item = await client.get('history', tweetId)
    return item ? DownloadHistoryEntity.build(item) : undefined
  }

  async searchByUserName(userName: string, limit = 10): Promise<DownloadHistoryEntity[]> {
    const client = await this.clientProvider()
    const items = await client.getAll('history')
    const lowerCaseName = userName.toLowerCase()
    return items
      .filter(
        item =>
          item.displayName.toLowerCase().includes(lowerCaseName) ||
          item.screenName.toLowerCase().includes(lowerCaseName)
      )
      .map(DownloadHistoryEntity.build)
      .splice(0, limit)
  }

  search(...predicates: DownloadItemPredicate[]) {
    return async (
      limit?: number,
      skip?: number
    ): Promise<[DownloadHistoryEntity[], number]> => {
      const collection = this.openCursor(null, 'prev', 'inf')('byDownloadTime')
      const result: DownloadHistoryEntity[] = []

      let matchedCount = 0
      const increaseCount = () => (matchedCount += 1)
      for await (const v of collection) {
        if (Array.from(predicates).every(predicate => predicate(v))) {
          if (skip) {
            skip -= 1
            increaseCount()
            continue
          }
          if (result.length < limit) result.push(DownloadHistoryEntity.build(v))
          increaseCount()
        }
      }

      return [result, matchedCount]
    }
  }
}

const asyncGeneratorToArray = async <T>(
  generator: AsyncGenerator<T, void, unknown>
): Promise<T[]> => {
  const result: T[] = []
  for await (const v of generator) result.push(v)
  return result
}

export interface ITweetHashtagRepository {
  getAllTags(): Promise<string[]>
  getAll(): Promise<HashtagItem[]>
  getTweetsByTag(tag: string): Promise<string[]>
  getTweetsByTags(...tags: string[]): Promise<string[]>
  addTweet(tweetId: string): (...tag: string[]) => Promise<void>
  clear(): Promise<void>
}

export class IndexedDBHashtagRepository
  extends IndexedDBRepository
  implements ITweetHashtagRepository
{
  async getAllTags(): Promise<string[]> {
    const client = await this.clientProvider()
    return client.getAllKeys('hashtag')
  }

  async getAll(): Promise<HashtagItem[]> {
    const client = await this.clientProvider()
    return client.getAll('hashtag')
  }

  async getTweetsByTag(tag: string): Promise<string[]> {
    const client = await this.clientProvider()
    const item = await client.get('hashtag', tag)
    return item ? Array.from(item.tweetIds) : []
  }

  async getTweetsByTags(...tags: string[]): Promise<string[]> {
    const ids = new Set<string>()
    const client = await this.clientProvider()
    for (const tag of tags) {
      const item = await client.get('hashtag', tag)
      item && item.tweetIds.forEach(id => ids.add(id))
    }
    return Array.from(ids)
  }

  addTweet(tweetId: string): (...tags: string[]) => Promise<void> {
    return async (...tags: string[]) => {
      if (tags.length === 0) return
      const client = await this.clientProvider()
      for (const tag of tags) {
        const item = await client.get('hashtag', tag)
        const ids = item?.tweetIds || new Set()
        await client.put('hashtag', {
          name: tag,
          tweetIds: ids.add(tweetId),
        })
      }
    }
  }

  async clear() {
    const client = await this.clientProvider()
    await client.clear('hashtag')
  }
}
