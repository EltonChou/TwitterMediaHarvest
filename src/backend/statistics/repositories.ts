import {
  BrowserStorageFetcher,
  BrowserStorageSetter,
  storageFetcher,
  storageSetter
} from '../../libs/chromeApi'

export enum StatisticsKey {
  SuccessDownloadCount = 'successDownloadCount',
  FailedDownloadCount = 'failedDownloadCount',
  ErrorCount = 'errorCount',
}

type DownloadStatistic = {
  [StatisticsKey.SuccessDownloadCount]?: number
  [StatisticsKey.FailedDownloadCount]?: number
  [StatisticsKey.ErrorCount]?: number
}

const defaultStatistic: DownloadStatistic = {
  [StatisticsKey.ErrorCount]: 0,
  [StatisticsKey.FailedDownloadCount]: 0,
  [StatisticsKey.SuccessDownloadCount]: 0,
}

export interface IStatisticsRepository {
  getStatisticsCount(key: StatisticsKey): Promise<number>
  addStatisticsCount(key: StatisticsKey): Promise<void>
}

export default class StatisticsRepository implements IStatisticsRepository {
  private fetchStorage: BrowserStorageFetcher
  private setStorage: BrowserStorageSetter

  constructor(storageArea: chrome.storage.StorageArea) {
    this.fetchStorage = storageFetcher(storageArea)
    this.setStorage = storageSetter(storageArea)
  }

  async getStatisticsCount(
    key: StatisticsKey
  ): Promise<number> {
    const downloadStatistic: DownloadStatistic = {}
    downloadStatistic[key] = 0

    const count = await this.fetchStorage(downloadStatistic)
    return count[key]
  }

  async addStatisticsCount(key: StatisticsKey): Promise<void> {
    const count = await this.getStatisticsCount(key)
    const downloadCount: DownloadStatistic = {}
    downloadCount[key] = count + 1

    await this.setStorage(downloadCount)
  }

  async setDefaultStatistics(): Promise<void> {
    await this.setStorage(defaultStatistic)
  }
}