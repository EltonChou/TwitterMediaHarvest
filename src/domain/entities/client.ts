import type { UsageStatistics } from '#domain/valueObjects/usageStatistics'
import { TimeHelper } from '#helpers/time'
import { Entity, EntityId } from './base'

type ClientProps = {
  syncToken: string
  uninstallCode: string
  syncedAt: number

  usageStatistics: UsageStatistics
}

export class ClientUUID extends EntityId<string> {}

const SYNC_PERIOD: number =
  process.env.NODE_ENV === 'production' ? TimeHelper.minute(30) : TimeHelper.minute(10)

/**
 * syncing period is 30 minutes in prodcution, otherwise 10 minutes.
 */
export class Client extends Entity<ClientUUID, ClientProps> {
  get syncToken() {
    return this.props.syncToken
  }

  get uninstallCode() {
    return this.props.uninstallCode
  }

  get shouldSync(): boolean {
    return Date.now() - this.props.syncedAt >= SYNC_PERIOD
  }

  get usageStatistics() {
    return this.props.usageStatistics
  }

  updateSyncToken(token: string) {
    this.props.syncToken = token
  }
}
