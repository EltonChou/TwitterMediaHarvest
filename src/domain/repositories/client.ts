import type { Client } from '#domain/entities/client'

export interface IClientRepository {
  get(): AsyncResult<Client>
  sync(client: Client): Promise<UnsafeTask>
}
