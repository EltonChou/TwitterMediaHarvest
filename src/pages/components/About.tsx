import React, { useEffect, useState } from 'react'

import { clientInfoRepo } from '@backend/configurations'
import { Link, Stack, Text } from '@chakra-ui/react'
import ExtLinks from '@pages/links'
import { i18n } from '@pages/utils'
import browser from 'webextension-polyfill'

const ProductInformation = () => {
  const [clientUuid, setClientUuid] = useState('xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx')

  useEffect(() => {
    clientInfoRepo.getInfo().then(info => {
      setClientUuid(info.uuid)
    })
  }, [])

  return (
    <Stack>
      <Text fontSize={'1.5em'}>{browser.runtime.getManifest().name}</Text>
      <Text fontSize={'sm'} color={'gray.500'}>
        {clientUuid}
      </Text>
      <Text color={'gray.400'}>
        {i18n('options_about_version')} {browser.runtime.getManifest().version_name}
      </Text>
    </Stack>
  )
}

const Links = () => {
  return (
    <Stack>
      <Link isExternal referrerPolicy="no-referrer" href={ExtLinks.website} textTransform={'capitalize'}>
        {i18n('options_about_officialWebsite')}
      </Link>
      <Link isExternal referrerPolicy="no-referrer" href={ExtLinks.privacy} textTransform={'capitalize'}>
        {i18n('options_about_privacyPolicy')}
      </Link>
      <Link isExternal referrerPolicy="no-referrer" href={ExtLinks.issues} textTransform={'capitalize'}>
        {i18n('options_about_issues')}
      </Link>
      <Link isExternal referrerPolicy="no-referrer" href={ExtLinks.changelog} textTransform={'capitalize'}>
        {i18n('options_about_changeLog')}
      </Link>
      <Link isExternal referrerPolicy="no-referrer" href={ExtLinks.github} textTransform={'capitalize'}>
        Github
      </Link>
    </Stack>
  )
}

const About = () => {
  return (
    <Stack p={'1.5rem'} spacing={10}>
      <ProductInformation />
      <Links />
    </Stack>
  )
}

export default About
