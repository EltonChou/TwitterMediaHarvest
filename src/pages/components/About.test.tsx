/**
 * @jest-environment jsdom
 */
import { MockClientRepository } from '#mocks/repositories/client'
import { toSuccessResult } from '#utils/result'
import { generateClient } from '#utils/test/client'
import About from './About'
import { render, screen } from '@testing-library/react'
import React from 'react'

describe('unit test for About component', () => {
  const clientRepo = new MockClientRepository()

  beforeAll(() => {
    jest.spyOn(clientRepo, 'get').mockResolvedValue(toSuccessResult(generateClient()))
  })

  afterAll(() => {
    jest.restoreAllMocks()
    jest.resetAllMocks()
  })

  it('can render properly', async () => {
    const { container, unmount } = render(<About clientRepo={clientRepo} />)

    const links = screen.getAllByTestId('information-link')

    expect(links).toHaveLength(5)
    expect(container).toMatchSnapshot()

    unmount()
  })
})
