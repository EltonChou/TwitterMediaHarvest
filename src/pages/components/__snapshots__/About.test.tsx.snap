// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit test for About component can render properly 1`] = `
<div>
  <div
    class="chakra-stack css-8osiqh"
  >
    <div
      class="chakra-stack css-8g8ihq"
    >
      <p
        class="chakra-text css-3fuzg5"
      />
      <div
        class="chakra-skeleton css-fotqef"
      >
        <p
          class="chakra-text css-5ft5pb"
        >
          xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
        </p>
      </div>
      <p
        class="chakra-text css-qljqz"
      >
        Translated&lt;options:about_Version&gt;
         
      </p>
    </div>
    <div
      class="chakra-stack css-8g8ihq"
      data-testid="information-links"
    >
      <a
        class="chakra-link css-10iahqc"
        data-testid="information-link"
        href="https://github.com/EltonChou/TwitterMediaHarvest"
        referrerpolicy="no-referrer"
        rel="noopener"
        target="_blank"
      >
        Translated&lt;options:about_Official website&gt;
      </a>
      <a
        class="chakra-link css-10iahqc"
        data-testid="information-link"
        href="https://github.com/EltonChou/TwitterMediaHarvest/blob/main/PRIVACY_POLICY.md"
        referrerpolicy="no-referrer"
        rel="noopener"
        target="_blank"
      >
        Translated&lt;options:about_Privacy policy&gt;
      </a>
      <a
        class="chakra-link css-10iahqc"
        data-testid="information-link"
        href="https://github.com/EltonChou/TwitterMediaHarvest/issues"
        referrerpolicy="no-referrer"
        rel="noopener"
        target="_blank"
      >
        Translated&lt;options:about_Reoprt issues&gt;
      </a>
      <a
        class="chakra-link css-10iahqc"
        data-testid="information-link"
        href="https://github.com/EltonChou/TwitterMediaHarvest/blob/main/CHANGELOG.md#undefined"
        referrerpolicy="no-referrer"
        rel="noopener"
        target="_blank"
      >
        Translated&lt;options:about_Changelog&gt;
      </a>
      <a
        class="chakra-link css-10iahqc"
        data-testid="information-link"
        href="https://github.com/EltonChou/TwitterMediaHarvest"
        referrerpolicy="no-referrer"
        rel="noopener"
        target="_blank"
      >
        Github
      </a>
    </div>
  </div>
</div>
`;
